### ISSUE TITLE
Interface pattern generates query plan with duplicate fetches to the same service

### ISSUE THREAD

**Issue by michael-watson** [apollographql/federation#294]
## Accounts Schema
```
type User @key(fields: "id") {
  id: ID!
  name: String
  username: String
}
```

## Products Service
```
type Book implements Product @key(fields: "isbn") {
  isbn: String!
  title: String
  year: Int
  name: String
  price: Int
}

type TV implements Product @key(fields: "id") {
  id: String!
  title: String
  year: Int
  name: String
  price: Int
}

type Computer implements Product @key(fields: "id") {
  id: String!
  title: String
  year: Int
  name: String
  price: Int
}

type Furniture implements Product @key(fields: "sku") {
  sku: String!
  name: String
  price: Int
  weight: Int
}

interface Product {
  name: String
  price: Int
}

extend type Query {
  topProducts: [Product]
}
```

## Reviews Service
```
extend type Book implements Product @key(fields: "isbn") {
  isbn: String! @external
  reviews: [Review]
}

extend type TV implements Product @key(fields: "id") {
  id: String! @external
  reviews: [Review]
}
extend type Computer implements Product @key(fields: "id") {
  id: String! @external
  reviews: [Review]
}

extend type Furniture implements Product @key(fields: "sku") {
  sku: String! @external
  reviews: [Review]
}

extend interface Product {
  reviews: [Review]
}

type Review @key(fields: "id") {
  id: ID!
  body: String
  author: User @provides(fields: "username")
  product: Product
}

extend type User @key(fields: "id") {
  id: ID! @external
  username: String @external
  reviews: [Review]
}
```

## Given this query
```
query ios_TopProducts {
  topProducts {
    name
    price
    reviews {
      author {
        name
        username
      }
      body
      id
    }
  }
}
```

## Generates a query plan with multiple fetches to the `topProducts.@.reviews.@.author` services. It would be a duplication of **n** where **n** is the number of types that implement the given interface.
```
QueryPlan {
  Sequence {
    Fetch(service: "products") {
      {
        topProducts {
          __typename
          ... on Book {
            name
            price
            __typename
            isbn
          }
          ... on Computer {
            name
            price
            __typename
            id
          }
          ... on Furniture {
            name
            price
            __typename
            sku
          }
          ... on TV {
            name
            price
            __typename
            id
          }
        }
      }
    },
    Flatten(path: "topProducts.@") {
      Fetch(service: "reviews") {
        {
          ... on Book {
            __typename
            isbn
          }
          ... on Computer {
            __typename
            id
          }
          ... on Furniture {
            __typename
            sku
          }
          ... on TV {
            __typename
            id
          }
        } =>
        {
          ... on Book {
            reviews {
              author {
                __typename
                id
                username
              }
              body
              id
            }
          }
          ... on Computer {
            reviews {
              author {
                __typename
                id
                username
              }
              body
              id
            }
          }
          ... on Furniture {
            reviews {
              author {
                __typename
                id
                username
              }
              body
              id
            }
          }
          ... on TV {
            reviews {
              author {
                __typename
                id
                username
              }
              body
              id
            }
          }
        }
      },
    },
    Parallel {
      Flatten(path: "topProducts.@.reviews.@.author") {
        Fetch(service: "accounts") {
          {
            ... on User {
              __typename
              id
            }
          } =>
          {
            ... on User {
              name
            }
          }
        },
      },
      Flatten(path: "topProducts.@.reviews.@.author") {
        Fetch(service: "accounts") {
          {
            ... on User {
              __typename
              id
            }
          } =>
          {
            ... on User {
              name
            }
          }
        },
      },
      Flatten(path: "topProducts.@.reviews.@.author") {
        Fetch(service: "accounts") {
          {
            ... on User {
              __typename
              id
            }
          } =>
          {
            ... on User {
              name
            }
          }
        },
      },
      Flatten(path: "topProducts.@.reviews.@.author") {
        Fetch(service: "accounts") {
          {
            ... on User {
              __typename
              id
            }
          } =>
          {
            ... on User {
              name
            }
          }
        },
      },
    },
  },
}
```

**Comment by ndintenfass** [comment]
Note form Solutions: not a blocker, but it's coming up every week with Walmart

**Comment by abernix** [comment]
The fix for this was released in `@apollo/gateway@0.27.1`! 🎉
